### PR TITLE
Renaming 'volumes' to 'files' and adding SecStr support

### PR DESCRIPTION
**Pull request by samuchila**
Rename of field 'volumes' to 'files' in service_config & companions. Being done in relation to [issue#8](https://github.com/aixigo/PREvant/issues/8)
- Service config and Companion has the ability to use both volumes and files in the JSON
- 'volumes' field to be deprecated, relevant messages added to openApi.yaml
- Change affects 11 files in total : Backward compatibility tests in place for serialization and deserialization of the new field files and old field volumes. No test cases written for internally called functions.
- File data is now stored as a SecStr to avoid data leakage within the program.

### CODE REVIEW

File: api/src/models/service_config/mod.rs [hunk 1]
```diff
@@ -339,6 +366,27 @@ mod tests {
     use secstr::SecUtf8;
     use serde_json::from_value;
 
+    fn desecure_files(
```
Comment (schrieveslaach) [inline api/src/models/service_config/mod.rs:369]: Please, remove. You can compare the results in the test by creating `SecStr` instead of `String` in the assertions.

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
Thanks for contributing this PR to PREvant. From the outside perspective the change look good to me because the are backward compatible. However, there are some internal changes that I would like to be improved (see comments).

Additionally, check all comments in the code base that mention #8. These comments should be resolved when merging and therefore, they can be removed with this PR.

File: api/src/models/service_config/mod.rs [hunk 2]
```diff
@@ -70,6 +77,26 @@ impl ServiceConfig {
         }
     }
 
+    fn deserialize_treemap<'de, D>(
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/models/service_config/mod.rs]: deserialize_treemap is no longer required as SecUtf8 has in built deserial capability.

File: api/src/infrastructure/kubernetes/payloads.rs [hunk 3]
```diff
@@ -163,8 +182,8 @@ pub fn deployment_payload(
         BTreeMap::from([(IMAGE_LABEL.to_string(), strategy.image().to_string())])
     };
 
-    let volume_mounts = strategy.volumes().map(|volumes| {
-        let parent_paths = volumes
+    let file_mounts = desecure_files(strategy.files().cloned()).map(|files| {
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/infrastructure/kubernetes/payloads.rs]: All instances of desecure_files removed.

File: api/src/infrastructure/kubernetes/infrastructure.rs [hunk 4]
```diff
@@ -341,8 +363,8 @@ impl KubernetesInfrastructure {
         strategy: &'a DeploymentStrategy,
         container_config: &ContainerConfig,
     ) -> Result<&'a DeploymentStrategy, KubernetesInfrastructureError> {
-        if let Some(volumes) = strategy.volumes() {
-            self.deploy_secret(app_name, strategy, volumes).await?;
+        if let Some(files) = KubernetesInfrastructure::desecure_files(strategy.files().cloned()) {
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/infrastructure/kubernetes/infrastructure.rs]: secrets_payload handles the conversion of SecUtf8 to String to create required V1Secret

File: api/src/config/mod.rs [hunk 5]
```diff
@@ -229,10 +228,10 @@ impl Service {
             for s in secrets.iter().filter(|s| s.matches_app_name(app_name)) {
                 let (path, sec) = s.clone().into();
 
-                service_config.add_volume(
+                service_config.add_file(
                     path,
                     // TODO: use secstr in service_config (see issue #8)
-                    String::from(sec.unsecure()),
+                    SecStr::from(sec.unsecure().as_bytes()),
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/config/mod.rs]: SecUtf8 to SecStr conversion also no longer required.

File: api/src/models/service_config/templating.rs [hunk 6]
```diff
@@ -97,6 +99,28 @@ impl ServiceConfig {
         self.apply_template(&parameters)
     }
 
+    fn desecure_files(
```
[the remark this answers is not in the document]
Comment (samuchila) [reply api/src/models/service_config/templating.rs]: Unsecure moved to render_template

File: api/src/apps/hooks.rs [hunk 7]
```diff
@@ -31,6 +31,7 @@ use boa_engine::property::Attribute;
 use boa_engine::syntax::ast::node::Node;
 use boa_engine::{Context, JsValue};
 use secstr::SecUtf8;
+use serde::Deserialize;
```
Comment (schrieveslaach) [inline api/src/apps/hooks.rs:34]: Unused? The new `use` statement in the tests mod seems to be unsed too. Please, check that

File: api/src/infrastructure/kubernetes/infrastructure.rs [hunk 8]
```diff
@@ -56,7 +56,6 @@ use std::convert::{From, TryFrom};
 use std::net::IpAddr;
 use std::path::PathBuf;
 use std::str::FromStr;
-
```
Comment (schrieveslaach) [inline api/src/infrastructure/kubernetes/infrastructure.rs:59]: Please, keep that line for better readability.

File: api/src/infrastructure/kubernetes/payloads.rs [hunk 9]
```diff
@@ -163,8 +162,8 @@ pub fn deployment_payload(
         BTreeMap::from([(IMAGE_LABEL.to_string(), strategy.image().to_string())])
     };
 
-    let volume_mounts = strategy.volumes().map(|volumes| {
-        let parent_paths = volumes
+    let file_mounts = strategy.files().map(|files| {
```
Comment (schrieveslaach) [inline api/src/infrastructure/kubernetes/payloads.rs:165]: Actually, I would say that `volume_mounts` is the correct name here because it maps to a Kubernetes primitive. 

```suggestion
    let volume_mounts = strategy.files().map(|files| {
``

File: api/src/infrastructure/kubernetes/payloads.rs [hunk 10]
```diff
@@ -179,13 +178,13 @@ pub fn deployment_payload(
             .collect::<Vec<_>>()
     });
 
-    let volumes = strategy.volumes().map(|volumes| {
-        let volumes = volumes
+    let files = strategy.files().map(|files| {
```
Comment (schrieveslaach) [inline api/src/infrastructure/kubernetes/payloads.rs:181]: See comment above: use K8s primitive.

```suggestion
    let volumes = strategy.files().map(|files| {
```

File: api/src/models/service_config/mod.rs [hunk 11]
```diff
@@ -26,23 +26,26 @@
 use crate::models::service::ContainerType;
 use crate::models::Image;
 pub use environment::{Environment, EnvironmentVariable};
+use secstr::SecUtf8;
 use serde::Deserialize;
 use serde_value::Value;
 use std::collections::BTreeMap;
 use std::hash::Hash;
 use std::path::PathBuf;
-
```
Comment (schrieveslaach) [inline api/src/models/service_config/mod.rs:34]: Please, keep that line.

File: api/src/models/service_config/mod.rs [hunk 12]
```diff
@@ -26,23 +26,26 @@
...
 mod environment;
 mod templating;
 
-#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq)]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
```
Comment (schrieveslaach) [inline api/src/models/service_config/mod.rs:38]: Why did you change that line?

File: api/src/models/service_config/mod.rs [hunk 13]
```diff
@@ -26,23 +26,26 @@
...
 #[serde(rename_all = "camelCase")]
 pub struct ServiceConfig {
     service_name: String,
     image: Image,
     env: Option<Environment>,
-    // TODO: rename this field because it does not match to volumes any more (it is file content, cf. issue #8)
-    volumes: Option<BTreeMap<PathBuf, String>>,
+    // Renamed volume field to files to add more relevance
+    // Functionality will remain as-is
+    // Files will now be stored with improved security using SecStr
```
Comment (schrieveslaach) [inline api/src/models/service_config/mod.rs:46]: I think this comment can be removed because that seems to be a note to yourself.

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
There are some minor things that should be worked out.

Additionally, please, provide a proper commit message that also then fixes #8 automatically.

File: api/src/apps/hooks.rs (hunk 7 above)
Comment (samuchila) [reply api/src/apps/hooks.rs]: There were a few repeated imports in tests which have been removed. The deserialize import also is removed as it is unused.

File: api/src/infrastructure/kubernetes/payloads.rs (hunk 9 above)
Comment (samuchila) [reply api/src/infrastructure/kubernetes/payloads.rs]: Followed the K8s primitives and reverted to previous version.

File: api/src/models/service_config/mod.rs (hunk 11 above)
Comment (samuchila) [reply api/src/models/service_config/mod.rs]: White space maintained for better readability.

File: api/src/models/service_config/mod.rs (hunk 13 above)
Comment (samuchila) [reply api/src/models/service_config/mod.rs]: Comment removed.

File: api/src/models/service_config/mod.rs (hunk 12 above)
Comment (samuchila) [reply api/src/models/service_config/mod.rs:39]: Hash is not implemented by SecUtf8. Also removing it did not change behavior. As discussed maintaining the derives without Hash

**Review by schrieveslaach — APPROVED** [review]
Look good to me. Thank you very much for your contribution.
